### ISSUE TITLE
HLSL: Casting rules for HLSL intrinsics

### ISSUE THREAD

**Issue by dankbaker** [KhronosGroup/glslang#449]
## Overivew of HLSL intrinsics

HLSL intrinsics are often thought of as functional calls that are built into the langauge. This is the natural assumption, coming from a C/C++ background. However, HLSL intrinsics have significantly different    casting rules then a C++ style does, some of which have no direct analogy. It's also important to note that HLSL does not use the same rules for function call casting as it does for intrinsic casting. This is mostly because the casting rules are not easy to express in a function template. I will attempt to document what I remember about casting rules, caveat emtpor that this could have changed a bit in the 10 years since I worked on it, and also my memory may be hazy in some areas.  I'm sure some of you know this info, but I wanted to have a more formal discussion of what the rules were as we know them.

I'll edit this as people point out flaws in my memory and/or updates the way the language is currently implemented.
## A brief history of why it's the way it is

The original HLSL came out 2002. At this time, GPUs were only capable of performing floating point arithmetic. Integers, bools, etc were simulated on top of floats. Thus, the int/uint/bool mechanics were somewhat syntactic sugar on top of a pipeline that was just floats. Later, these types became native but the old rules were still in play.
## The basic casting sequence for basic types and storage classes

An basic HLSL type essentially had three types, Scalar, Vector, and Matrix. Scalar always auto casts if needed, to a Matrix or Vector. A 1 dimensional Vector or matrix can auto cast down, if needed in most cases.

Further, in a situation where there are 2 or more types with different storage classes that must match, there is a casting sequences: 

bool <- uint <-- Int <-- Float <-- double

Not sure where halffloats go these days, need to experiments on that.  I think this part has changed a bit now and goes the other way. I'd have to run expirements, so don't quote me on this one ;)
## Matching an instrinsic

Intrinsics are matched on the a specific, (and undocumented, mind you), order. The first instrinsic where the arguments are deemed to match will be the one that is used. That is, there may be varients with different input types, in that case it would always go with the first one that matches. I don't believe you will see an ambiguity error for intrinsics, although that may have changed.  However, for most intrinsics this won't matter because there will really only be one possible choice.

IIRC, these things which are used for determining a successful match:
### storage classes

Like C, an Intrinsic is matched and casted based on it's argument list. An intrinsic has a list of arguments, each which have a vector class and an input class.  IIRC, the input classes would be: Bool, Int, UINT, float, Numeric (int, uint and float but bool), all (e.g all of the known types), and also various object types as appropriate.
### dimensional type

Each argument also can be marked as Scalar, Vector, Matrix. Invalid for objects, or maybe you could consider them scalars
### dimensional constraints

Arguments can have dimensional constraints associated with them. Easy example would be a cross product which has the constraint that the vectors must be 3 dimensional. More typically, constraints are not literal but cross-linked, which is discussed in next section.
### cross-linked/dependent typed arguments rule

This has no analogy in C, and there is no way to define a standalone HLSL function that does this. In addition to previous rules, arguments in HLSL intrinsics can be cross linked with one another. That is, Argument 2 could be declared to be the same storage and dimension as argument 1. This is very common for all the math instrinsics. For example, the  "min" intrinsic requires that argument 0 and 1 are the of the same type. Additionally, dimensional can be linked independent. For a "mul" case, argument 0 may require that the cols match the rows of arg 1 and vice versa.

When an intrinsic rule is evaluating,  We need to apply the constraints rule on the input argument list. This will either fail because the constraint is not possible, or it will create a 'union' type based on the type demotion rule. For example, for the instrinsic clamp(A,minA,maxA);  It would look at the types of A, minA, and maxA and find the least common demonator of the storage class. I can't recall how sizes are unioned, and it might have changed - I think that if one was scalar and was vector or matrix it would take the largest of the sizes, but someone should check that. 

After the this union of cross-linked types is performed (of the candidate input argument list), the argument list can then be validated against the acceptable input types provided for the intrinsic. 
### casting an instrinsic

Once an intrinsic has been accepted, that is the argument type list has been unioned according the rule set for that intrinsic,  you've got a unioned signature that could be a bit different then that actual input arguments. 

In this case, each argument that doesn't match would be issued the appropriate cast into this unioned type list. 
### TLDR

The cross-linking constraint rule is really the important concept, since it has no good analog in C. That is, the fact that HLSL examines different arguments from the argument list and generates a unioned type that is used to find the appropriate intrinsic and uses as the destination cast.

**Comment by johnkslang** [comment]
Thanks Dan.

I suspect the current method will work fine.  It is a two-layer thing.  The first layer is the set of intrinsic signatures in the symbol table for how they should operate internally, and the second layer is the policy for mapping call sites to signatures.  A lot can happen in the second layer, including intrinsic-specific policies, or even specific intrinsic-specific policies, worst case.

For example, in normal operation, if two arguments are "cross dependent", as in, say they must both be int or float, for some hypothetical intrinsic `foo()`, the first layer will put in the table:

```
foo(int, int)
foo(float, float)
```

Now, once that's done, all executions of foo will be one of those; with the second argument matching the first argument.  That should cover the cross-argument type rules.  The second layer will see a `foo()` call, go gather up the set of `foo` in the table and then make decisions like:

```
foo(2,3);       // exact match to foo(int,int), use that
foo(2.0, 4);    // ambiguous error, or pick foo(2.0,4.0)
foo(true,true); // pick foo(1, 1)
```

Those are example policies, the correct ones will be plugged in.  Normally, there is a set of weights for which argument conversions (and how many) are needed in a match, and if there is a clear singleton winner, that is used, while a tie results in an ambiguous-match error message.

Since the policies and weights can be as specific as needed, this design should work, and I suspect reasonably generically.

**Comment by dankbaker** [comment]
I don't believe you will ever be 100% compliant with this approach, though it may not matter much in practice. The current implementation appears to just make permutations for every storage class, rather then couple them during the name mangling process. So, at least currently the way it is done looks like it is pretty far off right now. Basically, looks to me like it's generating to many permutations for most intrinsics.

The question I have is if it's possible to create a set of casting weights that give you the same behavior. 
I'm not sure it is, though you can probably get pretty close.  I am less sure about vector size differences and casting aggregation, though I believe some of the more non standard things in HLSL that would fail this approach will issue warnings now on current HLSL compiler. The thing I am more wondering about is what happens with cases like:

```
int2 A;
float3 B;

float4 main() : SV_POSITION
{
    return dot(A,B);
}
```

which is a cast to int and dp2

vs 
```
int A;
float3 B;

float4 main() : SV_POSITION
{
    return dot(A,B);
}
```
which is a dp3 with a replicate

vs

```
int1   A;
float3 B;

float4 main() : SV_POSITION
{
    return dot(A,B);
}
```

Which ends up being a multiply. I haven't tried these cases in the GLSL front end, I think you could get these right with some casting weights, but might end up being a bit whack a moley from one intrinsic to the next.  I think we might end up in situations where the correct casting weights are different for different intrinsic.

The reality is that most of the  more esoteric rules won't come into play much, since it usually involves mixing types, dimensions and storage classes in ways that don't make much sense for most code.

**Comment by johnkslang** [comment]
That would actually all fall out naturally from my existing understanding of the implicit conversion rules:

- int -> float
- scalar -> vector (by smear)
- long vector -> shorter vector (by truncation)

Assuming layer 1 signatures of `foo(float, float)`, `foo(float2, float2)`, `foo(float3, float3)`, and `foo(float4, float4)`, followed by a layer 2 policy of simply try the standard implicit conversions I don't think would even find an ambiguity: each straightforwardly goes to the one you listed as the right result.  That is, this didn't even get into the space of multiple possible matches with weights selecting the best one, as there was only one choice for each anyway.

(These implicit conversion rules are not fully implemented yet though.) Exceptions can always be called out when we find them.

Also interesting to see the semantic difference shown here between int1 and int.

Great stuff, thanks.

**Comment by dankbaker** [comment]
I think you'd have to weight it still because otherwise you'd get the preference of case #2 reducing to a multiply because it would end up casting downward to a 1d vector and matching, even though you'd want to cast single channels up to a 3d vector. 
 
Thinking about it some more, I don't think it's necessary to worry about the ambiguous cases except to fail them rather then compile them. They aren't  common in HLSL and it's not unreasonable to have people specify. Usually the ambiguous cases  reducing a vector downward dimension ally, but I beleive even HLSL warns you these days if you do that.  

In fact, to start I'd think about failing cases where there is a vector truncation and it's amigous . since this is considered bad form in HLSL anyway (and as I mention, it does issue a warning).

**Comment by johnkslang** [comment]
FYI, I just posted a design for selection in issue #142.  For GLSL, the spec basically says a better single argument conversion is "float to double" beats others, and "int to float" beats "int to double".

For HLSL, I made a first stab, but very much welcome refinement.  Thanks.

**Comment by johnkslang** [comment]
This has been superseded by much work and learning. Remaining individual problems should be logged as  specific issues.
